### ISSUE TITLE
questions about waterlily

### ISSUE THREAD

**Issue by zrxdaly** [WaterLily-jl/WaterLily.jl#143]
Hi,
i am new to julia and waterlily. First, thanks for this beautiful solver. 
After trying some examples, i have some minor questions (maybe very basic), but i found hard to understand based on the documentation.
1. how can we define the grid size, this is the part i missed in the [simulation function](https://docs.juliahub.com/General/WaterLily/stable/#WaterLily.Simulation). I have tried to modify several inputs, but the L and H do not change in the gif I generated. 
2. I want to do i time average field and then output/plot it in the end. this is the example i had. i wonder what is the correct way of doing it. 

```
using WaterLily
include("TwoD_plots.jl")
function channel_flow(; Re = 360, stop = 1.)
    # Set simulation size & physical parameters
    n, m = 5 * 2^7, 2^7
    muv = 1.488e-5
    h = 1
    u_star = Re * muv / h
    g(i, t) = i == 1 ? (Re * muv)^ 2 / h^ 3 : 0

    sim = Simulation((n, m), (u_star, 0), h; ν=muv, g, perdir=(1,))

    # plot the profiles
    for t in 0:0.1:stop
        sim_step!(sim, t)
        @inside sim.flow.σ =+ sim.flow.u[:,1]
        println("tU/L=", round(t, digits=4))
    end
    contourf(sim.flow.σ, label="line plot", c=:blue)

end
channel_flow(; Re=360, stop=1.0)

```

**Comment by zrxdaly** [comment]
3. in the open channel case I am running, I would like the surface BC to be no-slip, lateral direction periodic. The top boundary condition for ux = 0, d (uy)/dz = 0. The lateral periodical BC is easy, but I couldn't figure out how to set other BCs.

**Comment by b-fg** [comment]
Hi! Thanks, and answering your questions:

1.  The first argument in `Simulation` is the grid size. So by increasing `n` and `m` in your example, you will be effectively increasing the resolution of the simulation. The gif does not really care about this, but note that `countourf` has a `size` keyword argument to increase the size of the rendered image. For example, with a small flow resolution and large figure size you would see a blurry image.
2. Time averaging is still not part of of the codebase, but there is a PR (https://github.com/WaterLily-jl/WaterLily.jl/pull/132) that shows how to do it. You can look at that and try to implement it yourself. Feel free to ask question regarding that PR.
3. There is currently no interface for custom BCs. I pointed this out to @marinlauber in the PR https://github.com/WaterLily-jl/WaterLily.jl/pull/136. This should be a future feature as well.

And just for clarity, based on the boundary conditions you mention, the simulation you are currently trying to set up is a boundary-layer flow, right?

**Comment by marinlauber** [comment]
Hi @zrxdaly, 

I had a quick scan of your script, and I have a few comments:

1. In WaterLily, we scale the velocity by the reference velocity to set the maximum flow velocity inside the domain to `maximum(sim.flow.u)~1` (this is done such that the time step becomes a CFL, as we have unit cells). This means that you have to set the numerical viscosity to `ν = U*L/Re` where `Re` is the Reynold number of your flow, `L` is the length-scale (number of cell per geometry length, essentially your resolution), and `U` is the velocity scale, which following the start of the comment is normally `U=1`. For example, if your physical setup is a pipe of Diameter 0.1 m with a mean flow velocity of 10m/s (assuming water with ν = 1e-6) your physical Re is 1e6. If you decide to have a pipe with 128 cells along the diameter, then `L=128` and `U=1`, you get a numerical viscosity of `ν = 1*128/1e6`. You can then scale your `g(i,t)` according to this new numerical viscosity and your pipe height `L`.

2. The line `@inside sim.flow.σ =+ sim.flow.u[:,1]` will not work, you should use something like `@inside sim.flow.σ[I] = sim.flow.u[I,1]`, or in your case, you can directly pass `sim.flow.u[:,:,1]` to `contour!`. Also, the assignment operator `=+` is not equivalent to `+=`, so make sure you are doing what you want. 

3. As @b-fg mentioned, there is no way of defining custom BCs (and probably won't be for a while). A good start will be to look at the [file](https://github.com/marinlauber/WaterLily/blob/4cc426cf19603f224c45432f1fec4e0b1bff71a3/examples/TwoD_LidCavity.jl), where I implemented a [lid-driven cavity](https://www.openfoam.com/documentation/tutorial-guide/2-incompressible-flow/2.1-lid-driven-cavity-flow). One thing to note is that we are using a staggered arrangement of the velocity and pressure, which means that some BCs have to be imposed through interpolation and cannot be imposed directly, see [here](https://github.com/marinlauber/WaterLily/blob/4cc426cf19603f224c45432f1fec4e0b1bff71a3/examples/TwoD_LidCavity.jl#L10). Lastly, the BC d(uy)/dz = 0. does not make sense in 2D flow, I suppose you mean d(uy)/dy=0? I can give you a hand with these BCs if that's not clear.

Marin

**Comment by zrxdaly** [comment]
Thank you two for the detailed answer. That indeed clarifies things. 
@b-fg , yes it is an idealized boundary layer flow. In the analogy to open channel flow, i would like to reproduce some typical channel flow profiles (wall-bounded flow). 

@marinlauber yes you are right. that is a typo. the top BCs should be d(ux)/dy = 0, uy = 0. And thanks for the nice lid-driven cavity case. I have modified the code based on your code. Coutours look fine and the profile also looks fine. 
![channel_flow](https://github.com/user-attachments/assets/057d609e-1507-44c6-934b-6e087a301fb3)

![plot_u](https://github.com/user-attachments/assets/49e144bb-989c-452b-8be9-efbf4e9314d7)

```
using WaterLily
include("TwoD_plots.jl")

mag(I, u) = √sum(ntuple(i -> 0.25 * (u[I, i] + u[I+δ(i, I), i])^2, length(I)))

function BC_CH!(a, A)
    N, n = WaterLily.size_u(a)
    for j ∈ 1:n, i ∈ 1:n
        if i == 1 && j == 2 # du/dy = 0; v = 0
            WaterLily.@loop a[I, i] =  a[I-δ(j, I), i] over I ∈ WaterLily.slice(N, N[j], j)
            WaterLily.@loop a[I, i] = -a[I+δ(j, I), i] over I ∈ WaterLily.slice(N, 1, j)
        elseif i == j # Normal direction, homoheneous Dirichlet
            WaterLily.@loop a[I, i] = 0.0 over I ∈ WaterLily.slice(N, 1, j)
            WaterLily.@loop a[I, i] = 0.0 over I ∈ WaterLily.slice(N, N[j], j)
        end
    end
end

@fastmath function WaterLily.mom_step!(a::Flow{N}, b::AbstractPoisson) where {N}
    a.u⁰ .= a.u
    WaterLily.scale_u!(a, 0)
    # predictor u → u'
    WaterLily.conv_diff!(a.f, a.u⁰, a.σ, ν=a.ν, perdir=a.perdir)
    WaterLily.BDIM!(a)
    BC_CH!(a.u, a.U)
    WaterLily.project!(a, b)
    BC_CH!(a.u, a.U)
    # corrector u → u¹
    WaterLily.conv_diff!(a.f, a.u, a.σ, ν=a.ν, perdir=a.perdir)
    WaterLily.BDIM!(a)
    WaterLily.scale_u!(a, 0.5)
    BC_CH!(a.u, a.U)
    WaterLily.project!(a, b, 0.5)
    BC_CH!(a.u, a.U)
    push!(a.Δt, WaterLily.CFL(a))
end

L = 2^5
U = 1.0
n, m = 5 * L, L
Re = 360
muv = U * L / Re
g(i, t) = i == 1 ? (Re * muv)^2 / L^3 : 0
sim = Simulation((n, m), (U, 0), L; ν=muv, g, perdir=(1,))

t₀ = round(sim_time(sim))
duration = 50;
step = 0.1;

anim = @animate for tᵢ in range(t₀, t₀ + duration; step)

    # update until time tᵢ in the background
    sim_step!(sim, tᵢ)

    # flood plot
    @inside sim.flow.σ[I] = mag(I, sim.flow.u)
    flood(sim.flow.σ; shift=(-0.5, -0.5), clims=(0, 1))

    # print time step
    println("tU/L=", round(tᵢ, digits=4), ", Δt=", round(sim.flow.Δt[end], digits=3))
end
gif(anim, "channel_flow.gif", fps=15)

```

**Comment by marinlauber** [comment]
@zrxdaly I ran our script and it worked but its not doing what you expect. In the 'mom_step!' function that you overwrite you need to call 'accelerate!' function, you can have a look in the 'src/Flow.jl' file for when it needs to be placed. In the lid cavity case I didn't add it since there is not body force, but in yours it's the forcing mechanism that drives the flow. 

Having said that, the flow still accelerates in the x direction and I think this is because you boundary conditions are incorrect and this sets a large divergence at the inlet/outlet and as a result a large pressure field that drives the flow. If you look at the pressure field you will see the problem 'flood(sim.flow.p)'. Additionally, you are never setting the periodic BCs in your 'BC!' function. You should look into 'src/util.jl' for the normal BC function and use that as a baseline for your script.

In your case (2D) I will not expect the boundary layer to become unstead, in fact you should get exactly a steady laminar boundary layer for sufficient convective time with this setup.

Marin

**Comment by marinlauber** [comment]
@zrxdaly I had a quick try and I think it should look something like this

```
using WaterLily
include("../src/TwoD_plots.jl")

# velocity magnitude
mag(I,u) = √sum(ntuple(i->0.25*(u[I,i]+u[I+δ(i,I),i])^2,length(I)))

# import explicitly BC function and overwrite
function BC_CH!(a,perdir=())
    N,n = WaterLily.size_u(a)
    for j ∈ 1:n, i ∈ 1:n
        if j in perdir
            @WaterLily.loop a[I,i] = a[WaterLily.CIj(j,I,N[j]-1),i] over I ∈ WaterLily.slice(N,1,j)
            @WaterLily.loop a[I,i] = a[WaterLily.CIj(j,I,2),i] over I ∈ WaterLily.slice(N,N[j],j)
        else
            if i==1 && j==2 # u-velocity on top and bottom boundaries
                @WaterLily.loop a[I,i] =  a[I-δ(j,I),i] over I ∈ WaterLily.slice(N,N[j],j) # d(ux)/dy = 0
                @WaterLily.loop a[I,i] = -a[I+δ(j,I),i] over I ∈ WaterLily.slice(N,1,j)    # ux = 0 on bottom (no slip), must interpolate
            else # v-velocity on top and bottom boundaries
                @WaterLily.loop a[I,i] = 0.0 over I ∈ WaterLily.slice(N,1,j)    # v = 0
                @WaterLily.loop a[I,i] = 0.0 over I ∈ WaterLily.slice(N,N[j],j) # v = 0, no slip
            end
        end
    end
end
@fastmath function WaterLily.mom_step!(a::Flow{N},b::AbstractPoisson) where N
    a.u⁰ .= a.u; WaterLily.scale_u!(a,0)
    # predictor u → u'
    WaterLily.conv_diff!(a.f,a.u⁰,a.σ,ν=a.ν,perdir=a.perdir)
    WaterLily.accelerate!(a.f,@view(a.Δt[1:end-1]),a.g,a.U)
    WaterLily.BDIM!(a); BC_CH!(a.u,a.perdir)
    WaterLily.project!(a,b); BC_CH!(a.u,a.perdir)
    # corrector u → u¹
    WaterLily.conv_diff!(a.f,a.u,a.σ,ν=a.ν,perdir=a.perdir)
    WaterLily.accelerate!(a.f,@view(a.Δt[1:end-1]),a.g,a.U)
    WaterLily.BDIM!(a); WaterLily.scale_u!(a,0.5); BC_CH!(a.u,a.perdir)
    WaterLily.project!(a,b,0.5); BC_CH!(a.u,a.perdir)
    push!(a.Δt,WaterLily.CFL(a))
end

# Initialize simulation
L = 2^7
U = 1.0
Re = 100
# pressure gradient required to drive the flow to u~1
g(i, t) = i == 1 ? U^2/(L/2)^2 : 0

# using CUDA
sim = Simulation((4L,L),(0.0,0.0),L;U=U,ν=U*L/Re,g,perdir=(1,),mem=Array)

# get start time
t₀ = round(sim_time(sim))
duration = 50; step = 0.1

anim = @animate for tᵢ in range(t₀,t₀+duration;step)

    # update until time tᵢ in the background
    sim_step!(sim,tᵢ)

    # flood plot
    @inside sim.flow.σ[I] = mag(I,sim.flow.u)
    flood(sim.flow.σ[inside(sim.flow.σ)]|>Array; shift=(-0.5,-0.5),clims=(0,1))

    # print time step
    println("tU/L=",round(tᵢ,digits=4),", Δt=",round(sim.flow.Δt[end],digits=3))
end
gif(anim,"chanel_flow.gif",fps=15)

# # plot the profile
# plot(sim.flow.u[:,2:end,1]'./maximum(sim.flow.u[:,:,1]),collect(1:129)./129,
#     label=:none,xlabel="u/U",ylabel="y/L")
# # analytical solution laminar chanel flow u/U ~ C*(y/L)-(y/L)^2 ∀ y ∈ [0,L/2]
# plot!(4*(collect(0:0.01:0.5).-collect(0:0.01:0.5).^2),2collect(0:0.01:0.5),label="analytical")
# savefig("chanel_flow.png")
```

This gets me almost the analytical profiles for a laminar channel flow (lower half) 
![chanel_flow](https://github.com/user-attachments/assets/d50394eb-f882-468a-a9c2-7960619132d7)

If you want to get a turbulent channel flow you will need to do 3D simulations on a bigger domain.

Marin

**Comment by zrxdaly** [comment]
@marinlauber thanks a lot for your effort and explanation. 
It seems that the periodic BC also needs to be reloaded as well.
I don't fully understand how BCs are indexed in Waterlily.   
1. if i see your code comments corretly, for `u`, `slice(N,N[j],j)` is the top BC and `slice(N,1,j)` is bottom BC. For `v`, it is another way around. How do i understand this, `slice` function does not give any clue to me. sorry. 
2. for both `u` and `v`, the bottom BC is 0. However, `u` is interpolated with `a[I,i] = -a[I+δ(j,I),i]` and `v` is done with `a[I,i] = 0.0`. how are these two different
Thanks again for your patience with these basic questions.
Dai

**Comment by marinlauber** [comment]
Hi @zrxdaly ,

It might be easier to explain this in person, and potentially clarify some other stuff. Send me an email and we can plan a meeting next week.

Marin

**Comment by b-fg** [comment]
I would also suggest to provide a summary of the meeting here, so it can also be useful for other users.

**Comment by zrxdaly** [comment]
Thanks @marinlauber for taking the time to meet with me. I summarized some takeaway points here:

1. the velocity components are located on different faces of the cell, then v (vertical component) can be directly assigned and u needs to be interpolated. 
2. taking one line as an example: `@WaterLily.loop a[I,i] = a[WaterLily.CIj(j,I,N[j]-1),i] over I ∈ WaterLily.slice(N,1,j)` waterlily will loop over velocity component a_i, i (1, 2) represent the two velocity components (3 in the case of 3D). I is the index given by the `slice` method. `slice` takes `N` (`n * m`) size of array, indexing at the dimention of j (in this case `j = 1`, since we set horizontal to be periodic). The second argument `1` means that we index at the most left side of the lateral direction. These indexes are replaced by the most right indexes given by `WaterLily.CIj(j,I,N[j]-1)` function.
